DataBuilder - support long names in files

rpalcolea opened the pull request.

rpalcolea commented:
@tcurdt this is for https://github.com/tcurdt/jdeb/issues/296

tcurdt commented on src/test/java/org/vafer/jdeb/DebMakerTestCase.java:
```diff
@@ -232,4 +241,180 @@ public void visit(TarArchiveEntry entry, byte[] content) throws IOException {
+            }
+        }, Compression.GZIP);
+
+        assertTrue("longname file wasn't found in the package", filesInDeb.containsKey(".//var/log/loooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongname/"));
```
The key should also use `longPathSuffix()` so that changing the `longPath` is one place.

tcurdt commented on src/test/java/org/vafer/jdeb/DebMakerTestCase.java:
```diff
@@ -232,4 +241,180 @@ public void visit(TarArchiveEntry entry, byte[] content) throws IOException {
+        assertTrue("Cannot delete the file " + deb, deb.delete());
+    }
+
+    private String longPathSuffix() {
```
I would generate a long path based on `numDirectories` being the number of directory levels and `nameLength` being the length of the file name. For now they could be just local variables. But testing a few combinations could be a could idea. Pseudo code as in:

```
test(longPath(numDirectories=2, nameLength=10))  == OK
test(longPath(numDirectories=5, nameLength=30))  == FAIL
test(longPath(numDirectories=5, nameLength=120)) == FAIL
```

tcurdt commented on src/test/java/org/vafer/jdeb/DebMakerTestCase.java:
```diff
@@ -232,4 +241,180 @@ public void visit(TarArchiveEntry entry, byte[] content) throws IOException {
+            }
+        }, Compression.GZIP);
+
+        assertTrue("longname file wasn't found in the package", filesInDeb.containsKey(".//var/log/loooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongnameloooongname/"));
```
The key should also use longPathSuffix() so that changing the longPath is one place.

tcurdt commented on src/test/java/org/vafer/jdeb/DebMakerTestCase.java:
```diff
@@ -232,4 +241,180 @@ public void visit(TarArchiveEntry entry, byte[] content) throws IOException {
+            }
+        }, Compression.GZIP);
+
+        assertTrue("testfile wasn't found in the package", filesInDeb.containsKey("./test/testfile"));
```
I don't quite understand the idea behind this test and the asserts.

tcurdt commented on src/test/java/org/vafer/jdeb/DebMakerTestCase.java:
```diff
@@ -232,4 +241,180 @@ public void visit(TarArchiveEntry entry, byte[] content) throws IOException {
+            }
+        }, Compression.GZIP);
+
+        assertTrue("testfile wasn't found in the package", filesInDeb.containsKey("./test/testfile"));
```
I don't quite understand the idea behind this test and the asserts.

tcurdt commented:
Would be great if you could adjust the test a bit. But thanks for the quick PR!

rpalcolea commented:
Hi @tcurdt , moved everything into a Parameterized test => https://github.com/tcurdt/jdeb/pull/297/commits/2e4a25bf3bace5ced37b0f7890ae61f65d8472ca

Also removed the couple of tests that didn't make sense

tcurdt commented:
Much better! AppVeyor (the windows build) is failing though. Maybe just a hiccup. Feel free to check or I will have to check tomorrow. But from on the first glance it looks good! Thanks!

rpalcolea commented:
@tcurdt , not very familiar with appveyor and windows. If you could take a look it would be phenomenal! thanks for your prompt responses

tcurdt commented:
So it reports
```
rg.vafer.jdeb.DebMakerLongNameTestCase
testLongLinkName[0](org.vafer.jdeb.DebMakerLongNameTestCase)  Time elapsed: 0.047 sec  <<< FAILURE!
java.lang.AssertionError: longname file wasn't found in the package
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.vafer.jdeb.DebMakerLongNameTestCase.testLongLinkName(DebMakerLongNameTestCase.java:89)
testLongLinkName[1](org.vafer.jdeb.DebMakerLongNameTestCase)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: longname file wasn't found in the package
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.vafer.jdeb.DebMakerLongNameTestCase.testLongLinkName(DebMakerLongNameTestCase.java:89)
testLongLinkName[2](org.vafer.jdeb.DebMakerLongNameTestCase)  Time elapsed: 0.015 sec  <<< FAILURE!
java.lang.AssertionError: longname file wasn't found in the package
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.vafer.jdeb.DebMakerLongNameTestCase.testLongLinkName(DebMakerLongNameTestCase.java:89)
testLongLinkName[3](org.vafer.jdeb.DebMakerLongNameTestCase)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: longname file wasn't found in the package
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.vafer.jdeb.DebMakerLongNameTestCase.testLongLinkName(DebMakerLongNameTestCase.java:89)
testLongLinkName[4](org.vafer.jdeb.DebMakerLongNameTestCase)  Time elapsed: 0.094 sec  <<< FAILURE!
java.lang.AssertionError: longname file wasn't found in the package
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.vafer.jdeb.DebMakerLongNameTestCase.testLongLinkName(DebMakerLongNameTestCase.java:89)
testLongLinkName[5](org.vafer.jdeb.DebMakerLongNameTestCase)  Time elapsed: 0.141 sec  <<< FAILURE!
java.lang.AssertionError: longname file wasn't found in the package
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.vafer.jdeb.DebMakerLongNameTestCase.testLongLinkName(DebMakerLongNameTestCase.java:89)
testLongLinkName[6](org.vafer.jdeb.DebMakerLongNameTestCase)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: longname file wasn't found in the package
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.vafer.jdeb.DebMakerLongNameTestCase.testLongLinkName(DebMakerLongNameTestCase.java:89)
testLongLinkName[7](org.vafer.jdeb.DebMakerLongNameTestCase)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: longname file wasn't found in the package
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.vafer.jdeb.DebMakerLongNameTestCase.testLongLinkName(DebMakerLongNameTestCase.java:89)
```

I would bet this is because of the `FileSeperator`.
I don't remember from the top of my head - but it might be still be `/` inside the archive.

rpalcolea commented:
ha good catch @tcurdt 

Fixed that, build is green now 👍

tcurdt commented:
Thanks!

rpalcolea commented:
No problem! any plans to do a release soon? or do you have other work in mind?

tcurdt commented:
I just looked through the open issues and PRs. Also updated the dependencies. Nothing really preventing a 1.8 release. I should be able to get it out in the next few days.

rpalcolea commented:
awesome thanks again !
